pmlc/dialect/stripe/util.cc

```diff
@@ -35,6 +37,166 @@ void createMainParallelFor(mlir::FuncOp funcOp) {
+    new_array.insert(new_array.begin(), elements.begin(), elements.end());
+  }
+  while (n >= new_array.size()) {
+    new_array.emplace_back(builder.getUnitAttr());
+  }
+  new_array[n] = elem;
+  ArrayRef new_array_ref(new_array.begin(), new_array.end());
+  return builder.getArrayAttr(new_array_ref);
+}
+
+void setOpAttrUnit(Operation* op, OpBuilder builder, const std::string& attr_name) {
+  if (!op) {
+    throw std::runtime_error("setUnitAttr: op is null");
+  }
+  auto old_attrs_dict = op->getAttrOfType<DictionaryAttr>(Dialect::getStripeAttrsName());
+  NamedAttribute new_elem = {builder.getIdentifier(attr_name), builder.getUnitAttr()};
+  auto new_attrs_dict = addAttrInDictionary(old_attrs_dict, builder, new_elem);
+  op->setAttr(Dialect::getStripeAttrsName(), new_attrs_dict);
+}
+
+void setIdxAttrUnit(ParallelForOp op, StringRef target_idx, const std::string& attr_name) {
+  auto idx_names = op.getAttrOfType<ArrayAttr>(mlir::Identifier::get("idx_names", op.getContext()));
+  if (!idx_names) {
+    return;
+  }
+  auto old_idx_attrs = op.getAttrOfType<ArrayAttr>(mlir::Identifier::get("idx_attrs", op.getContext()));
+  ArrayAttr new_idx_attrs;
+  auto builder = op.getBodyBuilder();
+  for (size_t i = 0; i < op.ranges().size(); i++) {
+    StringRef idx_name;
+    if (i < idx_names.size()) {
+      if (auto str_attr = idx_names.getValue()[i].dyn_cast<StringAttr>()) {
+        idx_name = str_attr.getValue();
+      }
+    }
+    if (idx_name == target_idx) {
+      DictionaryAttr old_dict;
+      if (old_idx_attrs && i < old_idx_attrs.size()) {
+        old_dict = old_idx_attrs.getValue()[i].dyn_cast<DictionaryAttr>();
+      }
+      NamedAttribute new_elem = {builder.getIdentifier(attr_name), builder.getUnitAttr()};
+      DictionaryAttr new_dict = addAttrInDictionary(old_dict, builder, new_elem);
+      new_idx_attrs = replaceAttrInArray(old_idx_attrs, builder, i, new_dict);
+      break;
+    }
+  }
+  op.setAttr(mlir::Identifier::get("idx_attrs", op.getContext()), new_idx_attrs);
+}
+
+int64_t idxRange(BlockArgument* idx) {
+  auto pf = mlir::cast<ParallelForOp>(idx->getOwner()->getParentOp());
+  return pf.getRange(idx->getArgNumber());
+}
+
+StringRef idxName(BlockArgument* idx) {
+  auto pf = mlir::cast<ParallelForOp>(idx->getOwner()->getParentOp());
+  auto names = pf.getAttrOfType<ArrayAttr>(mlir::Identifier::get("idx_names", pf.getContext()));
+  auto n = idx->getArgNumber();
+  auto idx_name = StringAttr::get("", pf.getContext());
+  if (names && n < names.size()) {
+    if (auto str_attr = names.getValue()[n].template dyn_cast<StringAttr>()) {
+      idx_name = str_attr;
+    }
+  }
+  return idx_name.getValue();
+}
+
+std::pair<StringRef, size_t> getSingleIndex(ParallelForOp op, size_t n) {
+  auto names = op.getAttrOfType<ArrayAttr>(mlir::Identifier::get("idx_names", op.getContext()));
+  auto ranges = op.ranges();
+  auto idx_name = StringAttr::get("", op.getContext());
```

Comment:
I'm thinking that in the case of an index name missing, we should generate a default one like `i{{n}}` or something.

Reply:
Also, it'd be nice to have a bunch of other code that does this same logic reuse this utility function. There's places in the code already that do this.